File: COURSE 2 (Data Structures)/week 1+week 2/Assignment/Submitted Files C++/check_brackets.cpp
#include <iostream>
#include <stack>
#include <string>
#include <algorithm>
using namespace std;
struct Bracket {
    Bracket(char type, int position):
        type(type),
        position(position)
    {}

    bool Matchc(char c) {
        if (type == '[' && c == ']')
            return true;
        if (type == '{' && c == '}')
            return true;
        if (type == '(' && c == ')')
            return true;
        return false;
    }

    char type;
    int position;
};

int main() {
    std::string text;
    getline(std::cin, text);

    std::stack <Bracket> opening_brackets_stack;
    int c=0;
    for (int position = 0; position < text.length(); ++position) {
        char next = text[position];

        if (next == '(' || next == '[' || next == '{') {
            Bracket *b=new Bracket(next,position+1);
            opening_brackets_stack.push(*b);
        }

        if (next == ')' || next == ']' || next == '}') {
            if(opening_brackets_stack.empty()==true || opening_brackets_stack.top().Matchc(next)==false){
                opening_brackets_stack.pop();
                c=1;
                cout<<position+1<<endl;
                break;
            }
            opening_brackets_stack.pop();
        }
    }

    if(c==0 && opening_brackets_stack.empty()==true){
        cout<<"Success"<<endl;
    }else if(c==0){
        cout<<opening_brackets_stack.top().position<<endl;
    }

    return 0;
}

File: COURSE 2 (Data Structures)/week 1+week 2/Assignment/Submitted Files Java/check_brackets.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.util.Stack;

class Bracket {

    char type;
    int position;

    Bracket(char type, int position) {
        this.type = type;
        this.position = position;
    }
}

class check_brackets {
    public static void main(String[] args) throws IOException {
        InputStreamReader input_stream = new InputStreamReader(System.in);
        BufferedReader reader = new BufferedReader(input_stream);
        String text = reader.readLine();

        Stack<Bracket> opening_brackets_stack = new Stack<>();
        int c = 1;
        for (int position = 0; position < text.length(); ++position) {
            char next = text.charAt(position);

            if (next == '(' || next == '[' || next == '{') {
                opening_brackets_stack.push(new Bracket(next, position + 1));
            }

            if (next == ')') {
                if (opening_brackets_stack.empty() || opening_brackets_stack.pop().type != '(') {
                    System.out.println(position + 1);
                    c = 0;
                    break;
                }
            }
            if (next == '}') {
                if (opening_brackets_stack.empty() || opening_brackets_stack.pop().type != '{') {
                    System.out.println(position + 1);
                    c = 0;
                    break;
                }
            }
            if (next == ']') {
                if (opening_brackets_stack.empty() || opening_brackets_stack.pop().type != '[') {
                    System.out.println(position + 1);
                    c = 0;
                    break;
                }
            }
        }

        if (c == 1 && opening_brackets_stack.empty()) {
            System.out.println("Success");
        } else if (c == 1) {
            System.out.println(opening_brackets_stack.pop().position);
        }
    }
}

